File: healthmate-frontend/src/services/follow.service.js
import api from './api';

const followTrainer = (id) => {
    return api.post(`/follow/trainer/${id}`);
};

const unfollowTrainer = (id) => {
    return api.delete(`/follow/trainer/${id}`);
};

const followUser = (id) => {
    return api.post(`/follow/user/${id}`);
};

const unfollowUser = (id) => {
    return api.delete(`/follow/user/${id}`);
};

const followCategory = (name) => {
    return api.post(`/follow/category/${name}`);
};

const unfollowCategory = (name) => {
    return api.delete(`/follow/category/${name}`);
};

const getMyFollows = () => {
    return api.get('/follow/my-follows');
};

const getMyFollowsResolved = () => {
    return api.get('/follow/my-follows/resolved');
};

const checkFollowing = (targetId, type) => {
    return api.get(`/follow/check?targetId=${targetId}&type=${type}`);
};

const getFollowerCount = (targetId, type = 'TRAINER') => {
    return api.get(`/follow/${targetId}/followers/count?type=${type}`);
};

const getFollowers = (targetId, type = 'TRAINER') => {
    return api.get(`/follow/${targetId}/followers?type=${type}`);
};

const getFollowingList = (followerId, type = 'TRAINER') => {
    return api.get(`/follow/${followerId}/following?type=${type}`);
};

const followService = {
    followTrainer,
    unfollowTrainer,
    followCategory,
    unfollowCategory,
    getMyFollows,
    checkFollowing,
    getFollowerCount,
    getFollowers,
    getFollowingList,
    followUser,
    unfollowUser,
    getMyFollowsResolved
};

export default followService;

File: healthmate-backend/src/main/java/com/healthmate/service/FollowService.java
package com.healthmate.service;

import com.healthmate.dto.FollowUserDTO;
import com.healthmate.enums.FollowType;
import com.healthmate.model.Follow;
import com.healthmate.repository.FollowRepository;
import com.healthmate.repository.UserRepository;
import com.healthmate.model.User;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class FollowService {

    @Autowired
    private FollowRepository followRepository;

    @Autowired
    private UserRepository userRepository;

    public void follow(String followerId, String targetId, FollowType type) {
        if (!followRepository.existsByFollowerIdAndTargetIdAndType(followerId, targetId, type)) {
            followRepository.save(new Follow(followerId, targetId, type));
        }
    }

    public void unfollow(String followerId, String targetId, FollowType type) {
        followRepository.deleteByFollowerIdAndTargetIdAndType(followerId, targetId, type);
    }

    public List<Follow> getMyFollows(String followerId) {
        return followRepository.findByFollowerId(followerId);
    }

    public boolean isFollowing(String followerId, String targetId, FollowType type) {
        return followRepository.existsByFollowerIdAndTargetIdAndType(followerId, targetId, type);
    }

    public long getFollowerCount(String targetId, FollowType type) {
        return followRepository.countByTargetIdAndType(targetId, type);
    }

    public List<FollowUserDTO> getFollowers(String targetId, FollowType type) {
        List<Follow> follows = followRepository.findByTargetIdAndType(targetId, type);
        return follows.stream()
                .map(f -> userRepository.findById(f.getFollowerId())
                        .map(u -> new FollowUserDTO(u.getId(), u.getUsername()))
                        .orElse(new FollowUserDTO(f.getFollowerId(), "Unknown User")))
                .collect(Collectors.toList());
    }

    public List<FollowUserDTO> getFollowing(String followerId, FollowType type) {
        List<Follow> follows = followRepository.findByFollowerIdAndType(followerId, type);
        return follows.stream()
                .map(f -> {
                    if (type == FollowType.TRAINER || type == FollowType.USER) {
                        return userRepository.findById(f.getTargetId())
                                .map(u -> new FollowUserDTO(u.getId(), u.getUsername()))
                                .orElse(new FollowUserDTO(f.getTargetId(), "Unknown User"));
                    } else {
                        return new FollowUserDTO(null, f.getTargetId()); // For CATEGORY, targetId is the name
                    }
                })
                .collect(Collectors.toList());
    }

    public List<FollowUserDTO> getAllFollowing(String followerId) {
        List<Follow> follows = followRepository.findByFollowerId(followerId);
        return follows.stream()
                .map(f -> {
                    if (f.getType() == FollowType.TRAINER || f.getType() == FollowType.USER) {
                        return userRepository.findById(f.getTargetId())
                                .map(u -> new FollowUserDTO(u.getId(), u.getUsername()))
                                .orElse(new FollowUserDTO(f.getTargetId(), "Unknown User"));
                    } else {
                        return new FollowUserDTO(null, f.getTargetId());
                    }
                })
                .collect(Collectors.toList());
    }
}
